src/common/functions/src/aggregates/aggregate_distinct_state.rs

```diff
@@ -0,0 +1,394 @@
+        validity: Option<&Bitmap>,
+        input_rows: usize,
+    ) -> Result<()> {
+        for row in 0..input_rows {
+            let values = columns.iter().map(|s| s.get(row)).collect::<Vec<_>>();
+            match validity {
+                Some(v) => {
+                    if v.get_bit(row) {
+                        let data_values = DataGroupValues(
+                            values
+                                .iter()
+                                .map(DataGroupValue::try_from)
+                                .collect::<Result<Vec<_>>>()?,
+                        );
+                        self.set.insert(data_values);
+                    }
+                }
+                None => {
+                    let data_values = DataGroupValues(
+                        values
+                            .iter()
+                            .map(DataGroupValue::try_from)
+                            .collect::<Result<Vec<_>>>()?,
+                    );
+                    self.set.insert(data_values);
+                }
+            }
+        }
+        Ok(())
+    }
+    fn merge(&mut self, rhs: &Self) -> Result<()> {
+        self.set.extend(rhs.set.clone());
+        Ok(())
+    }
+
+    fn build_columns(&self, fields: &[DataField]) -> Result<Vec<ColumnRef>> {
+        let mut results = Vec::with_capacity(self.set.len());
+        self.set.iter().for_each(|group_values| {
+            let mut v = Vec::with_capacity(group_values.0.len());
+            group_values.0.iter().for_each(|group_value| {
+                v.push(DataValue::from(group_value));
+            });
+
+            results.push(v);
+        });
+        let results = (0..fields.len())
+            .map(|i| {
+                results
+                    .iter()
+                    .map(|inner| inner[i].clone())
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
+        results
+            .iter()
+            .enumerate()
+            .map(|(i, v)| {
+                let data_type = fields[i].data_type();
+                data_type.create_column(v)
+            })
+            .collect::<Result<Vec<_>>>()
+    }
+}
+
+impl<T> DistinctStateFunc<T> for AggregateDistinctIntegerState<T>
+where T: PrimitiveType + std::hash::Hash + Eq + DFTryFrom<DataValue>
+{
+    fn new() -> Self {
+        AggregateDistinctIntegerState {
+            set: HashSet::new(),
+        }
+    }
+
+    fn serialize(&self, writer: &mut BytesMut) -> Result<()> {
+        serialize_into_buf(writer, &self.set)
+    }
+
+    fn deserialize(&mut self, reader: &mut &[u8]) -> Result<()> {
+        self.set = deserialize_from_slice(reader)?;
+        Ok(())
+    }
+
+    fn is_empty(&self) -> bool {
+        self.set.is_empty()
+    }
+
+    fn len(&self) -> usize {
+        self.set.len()
+    }
+
+    fn add(&mut self, columns: &[ColumnRef], row: usize) -> Result<()> {
+        let value = columns.get(0).map(|s| s.get(row)).unwrap();
```

Comment:
Need downcast the `columns[0]` to `PrimitiveColumn<T>`.

Reply:
`s.get(row)` is slow path.

Reply:
🤔 I feel that using value_unchecked does not improve performance significantly

Reply:
Yes, but if we downcast the `columns[0]` to `PrimitiveColumn<T>`, we can avoid the virtual call.